File: node/src/controller/pipefy.ts
import consumable from "../consumable/pipefy.js";
import external from "../external/pipefy.js";
import { data } from "../mockData/pipefy.js";

const iteration = async () => {
  const prospects = [...(await getProspects(1)), ...(await getProspects(2))];

  try {
    external.post("", prospects);
  } catch (err) {
    console.log("Erro no post dos prospects");
    console.log(err);
  }
};

const getProspects = async (tipoUsuario: 1 | 2) => {
  const req = await consumable(tipoUsuario);
  if (!req) return;

  const insertions = req.data.data.allCards.edges.reduce((acc: any) => {
    const node: any = req.data.data.allCards.edges[0].node;
    const mockData: any = data;
    const phaseName:
      | "Lead"
      | "Oportunidade"
      | "Cliente Adquirido"
      | "Cliente Perdido" = node.current_phase.name;

    const insertion = {
      idCliente: node.id,
      status: {
        Lead: 1,
        Oportunidade: 2,
        "Cliente Adquirido": 3,
        "Cliente Perdido": 4,
      }[phaseName],
      tipoUsuario,
    };

    for (let i = 0; i < node.fields.length; i++) {
      const fieldName = node.fields[i];
      const fieldData = mockData[fieldName.name] || null;

      if (!fieldData) continue;

      const optsValue = fieldData.opts?.[fieldName.value];
      const formatValue = fieldData.format?.(fieldName.value);
      const fieldValue = fieldName.value;
      // @ts-ignore
      insertion[fieldData.dbCol] =
        optsValue === undefined
          ? formatValue === undefined
            ? fieldValue
            : formatValue
          : optsValue;
    }
    return [...acc, insertion];
  }, []);

  return insertions;
};

export default iteration;

File: go/controller/pipefy.go
package controller

import (
	"fmt"
	"manuall/routine-api/consumable"
	"manuall/routine-api/external"
	"manuall/routine-api/mockData"
	"strconv"
)

func Pipefy() {
	executarTipo(1)
	executarTipo(2)
}

func executarTipo(tipo int) {

	graphqlResponse := consumable.Pipefy()

	var prospects []map[string]interface{}

	for _, edge := range graphqlResponse.Data.AllCards.Edges {
		resultado, err := strconv.Atoi(edge.Node.ID)
		if err != nil {
			fmt.Printf("Erro ao converter ID: %v\n", err)
			return
		}

		prospectAtual := map[string]interface{}{
			"idCliente":   resultado,
			"status":      mockData.SyncField("status").Opts[edge.Node.CurrentPhase.Name].(int),
			"tipoUsuario": tipo,
		}

		for _, field := range edge.Node.Fields {
			fieldConfig := mockData.SyncField(field.Name)
			var valor interface{} = nil

			if fieldConfig.Opts != nil {
				valor = fieldConfig.Opts[field.Value]
			} else {
				valor = field.Value
			}

			if fieldConfig.Format != nil {
				valor = fieldConfig.Format(valor.(string))
			}

			prospectAtual[fieldConfig.DbCol] = valor
		}

		prospects = append(prospects, prospectAtual)
	}

	external.PostarProspects(prospects)
}
